### DISCUSSION TITLE
Unexpected behavior surrounding transaction contexts

### DISCUSSION THREAD

**Discussion by filipemir** [psycopg/psycopg#826] [Q&A] [answered]
I came across some unexpected behavior of transaction contexts, and I'm looking for some help figuring out wether it's a result of a misunderstanding of mine or an actual bug. 

I work on an application that uses psycopg3 (v3.1.13). We have a DAO class with an update method, that I have simplified below:

```py
    async def update(
        self,
        id: int,
        value: str,
    ) -> Record:
        async with self.connection.transaction():
            query = sql.SQL(
                f"""
                    UPDATE record
                    SET column = %(value)s
                    WHERE id = %(id)s
                    RETURNING *;
                """
            )
            async with self.connection.cursor(
                row_factory=class_row(Record)
            ) as cursor:
                await cursor.execute(
                    query,
                    params={
                        "id": id,
                        "value": value,
                    },
                )

                return result
```

Where `self.connection` is an `AsyncConnection` instance.

(Previously that DAO method included two separate queries, hence the transaction, but I've removed the additional query for the sake of simplicity. The issue reproduces either way.)

The issue I'm seeing is that the update does not seem to commit immediately after we exit the transaction context. Instead it appears to only commit 200 to 300ms later, and as a result it isn't available to other connections until after that time is elapsed. Which is enough of a delay to cause issues in our use case.

This small change does resolve the issue for us, making it so that the change is committed by the time the method returns:
```diff
         id: int,
         value: str,
     ) -> Record:
-        async with self.connection.transaction():
+        async with self.connection.cursor():
             query = sql.SQL(
                 f"""
                     UPDATE record
@@ -139,6 +139,8 @@ class Dao:
                 )
                 result = await cursor.fetchone()

+                await self.connection.commit()
+
                 return result
```

What's puzzling me is that my reading of [the docs](https://www.psycopg.org/psycopg3/docs/basic/transactions.html#transaction-contexts) leads me to believe that the transaction should commit immediately as we exit the context:

> When the context is entered, a transaction is started; when leaving the context the transaction is committed, or it is rolled back if an exception is raised inside the block.

I guess I'm reading into that also that the commit is awaited. But that does not appear to be the case, in our usage at least. Is that expected? I can reproduce the issue reliably in our application. I have tried to create a simple reproducible example outside of our application, but so far I have not been able to. 

In case it matters, the nature of the `self.connection` doesn't seem to terribly matter. We use a connection pool but even just exposing a simple connection results in the same issue:

What am I missing?

**Comment by dvarrazzo** [comment, marked as answer]
I don't know what you are missing, the transaction block should works the
way you understand it.

One thing that I have in mind is that you might already be in a transaction
when you start the transaction() block. This would result in a nested
transaction, so at the end of the block you would only get a DISCARD
SAVEPOINT, not a COMMIT.

Can you check what is the `connection.info.transaction_status` right befor
the transaction() block? If it's INTRANS it's what I suppose it is. If it's
IDLE then there's something strange.

If you are not in a nested transaction then you'll have to debug yourself
what is happening, because your understanding of the transaction() block
behaviour is correct.

**Comment by filipemir** [comment]
@dvarrazzo thanks for the very quick reply. That was it. I had several preceding select statements that were initiating transactions as explained [here](https://www.psycopg.org/psycopg3/docs/basic/transactions.html):
> By default even a simple SELECT will start a transaction: in long-running programs, if no further action is taken, the session will remain idle in transaction, an undesirable condition for several reasons (locks are held by the session, tables bloat…). For long lived scripts, either make sure to terminate a transaction as soon as possible or use an autocommit connection.

Individually adding a `connection.commit()` after all of those queries fixes the problem. But this is making me realize that we are essentially developing under the assumption that `autocommit` was set to `True`. Calling `await connection.set_autocommit(True)` before exposing the connection for use also fixes the issue. So that's what we'll do. 

There is one detail that I still don't understand: why were the data mutations being committed at all? I see that the changes are available a few ms later, and yet I'm not explicitly committing any of the preceding or subsequent `SELECT` statements (or any subsequent writes). When i look in the DB logs I do indeed see a number of `BEGIN` statements but there is no `COMMIT` following the writes. A few milliseconds later I see a cascade of commits, one for each select I imagine. But I'm not explicitly committing anything. What else triggers the commit then?

**Comment by filipemir** [comment]
One point of clarification: the preceding `SELECT` statements weren't nested within the same connection context. They are all sequential. I.e. something like this:

```py
records = await dao.find_all(ids=[1,2,3])
await dao.update(id=1, value="foo")
```

Where the update method is the one above and the `find_all` method is something like:

```py
    async def find_all(
        self, ids: list[int]
    ) -> list[Record]:
        query = sql.SQL(
            """
                SELECT * from some_table
                WHERE id = ANY(%(ids)s);
            """
        )

        async with self.connection.cursor(
            row_factory=class_row(Record)
        ) as cursor:
            await cursor.execute(
                query, params={"ids": ids}
            )
            results = await cursor.fetchall()
            return results
```


My reading of the[ docs](https://www.psycopg.org/psycopg3/docs/basic/usage.html#with-connection) is that, because we are using a connection context there, the transaction wrapping the `SELECT` should commit once we exit that context, which should be before we call the `update` method. But that does not seem to happen in our application.

**Comment by filipemir** [comment]
In case it makes things a bit more concrete, I was able to improve my little reproduction to the point where it might be useful. Here it is: https://github.com/filipemir/psycopg-sandbox/blob/main/src/issue.py

It does still differ from what we see in our actual application in one major way: if I don't explicitly commit the transaction, it never gets committed. As I mentioned earlier, in our app we do see it getting committed, just a few milliseconds later. I wonder if this might have something to do with the difference in how we expose the connection. In our app we do:
```py
async def get_db_connection() -> AsyncGenerator[AsyncConnection[Any], None]:
    async with await AsyncConnection.connect(config.DATABASE_URL) as connection:
        yield connection
```

I couldn't get that to work in my reproduction, where we use:
```py
async def get_connection():
    con = await AsyncConnection.connect(
        "postgres://postgres:postgres@db:5432/?sslmode=disable"
    )
    return con
```
Which seems relevant to [this bit of the docs](https://www.psycopg.org/psycopg3/docs/advanced/async.html#with-async-connections)
